styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -127,38 +135,110 @@
   private final Stats stats;
   private final int pollPodsIntervalSeconds;
   private final ServiceAccountKeyManager serviceAccountKeyManager;
+  private final Supplier<String> secretEpochProvider;
 
   private Watch watch;
 
   private final Object secretMutationLock = new Object() {};
 
   KubernetesDockerRunner(NamespacedKubernetesClient client, StateManager stateManager, Stats stats,
-      ServiceAccountKeyManager serviceAccountKeyManager, int pollPodsIntervalSeconds) {
+      ServiceAccountKeyManager serviceAccountKeyManager, int pollPodsIntervalSeconds,
+      Supplier<String> secretEpochProvider) {
     this.stateManager = Objects.requireNonNull(stateManager);
     this.client = Objects.requireNonNull(client).inNamespace(NAMESPACE);
     this.stats = Objects.requireNonNull(stats);
     this.pollPodsIntervalSeconds = pollPodsIntervalSeconds;
-    this.serviceAccountKeyManager = serviceAccountKeyManager;
+    this.serviceAccountKeyManager = Objects.requireNonNull(serviceAccountKeyManager);
+    this.secretEpochProvider = Objects.requireNonNull(secretEpochProvider);
   }
 
   KubernetesDockerRunner(NamespacedKubernetesClient client, StateManager stateManager, Stats stats,
       ServiceAccountKeyManager serviceAccountKeyManager) {
-    this(client, stateManager, stats, serviceAccountKeyManager, DEFAULT_POLL_PODS_INTERVAL_SECONDS);
+    this(client, stateManager, stats, serviceAccountKeyManager, DEFAULT_POLL_PODS_INTERVAL_SECONDS,
+        DEFAULT_SECRET_EPOCH_PROVIDER);
   }
 
   @Override
   public String start(WorkflowInstance workflowInstance, RunSpec runSpec) throws IOException {
-    ensureSecrets(workflowInstance, runSpec);
+    final String secretEpoch = secretEpochProvider.get();
+    synchronized (secretMutationLock) {
+      ensureSecrets(workflowInstance, runSpec, secretEpoch);
+      try {
+        final Pod pod = client.pods().create(createPod(workflowInstance, runSpec, secretEpoch));
+        return pod.getMetadata().getName();
+      } catch (KubernetesClientException kce) {
+        throw new IOException("Failed to create Kubernetes pod", kce);
+      }
+    }
+  }
+
+  @Override
+  public void cleanup() throws IOException {
```

Comment:
How about splitting this method into three methods with comprehensive names instead of commenting?
1. Get all active secrets
2. Get all inactive secrets
3. Delete